posts/2020/03/pytorch-tensoriterator-internals.md

```diff
@@ -0,0 +1,236 @@
+data types.
+
+# Performing iterations
+
+Broadly, iterations using `TensorIterator` can be classified as point-wise iterations
+or reduction iterations. This plays a fundamental role in how iterations using `TensorIterator`
+are parallelized - point-wise iterations can be freely parallelized along any dimension
+and grain size while reduction operations have to be either parallelized along dimensions
+that you're not iterating over or by performing bisect and reduce operations along the
+dimension being iterated. Parallelization can also happen using vectorized operations.
+
+## Iteration details
+
+The simplest iteration operation can be performed using the 
+[`for_each`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L525) 
+function. This function has two overloads - one which accepts a loop of type `loop_t`
+and another which accepts a `loop2d_t` (find them [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.h#L166)). The former can iterate over a loop
+of a single dimension whereas the latter can do so over two dimensions. The simplest
+way of using `loop_t` is to pass it as a lambda using to `for_each`. A code snippet
+using it this way would look like so:
+
+``` cpp
+auto iter = TensorIterator();
+iter.add_output(out);
+iter.add_input(a);
+iter.build();
+
+auto loop = [&](char **data, const int64_t* strides, int64_t n) {
+    auto * out_data_bytes = data[0];
+    auto * in_data_bytes = data[1];
+    
+    // do something with input and output.
+    
+    out_data_bytes += strides[0];
+    in_data_bytes += strides[1];
+}
+
+iter.for_each(loop);
+```
+In the above example, the `char** data` gives a pointer to the data within the
+tensor in the same order that you specify when you build the iterator. Note
+that in order to make the implementation agnostic of any particular data type, you
+will always receive the data typecast to `char` (think of it as a bunch of bytes).
+
+The second argument is `int64_t* strides` which is an array containing the strides of
+the dimension that you're iterating over. We can add this stride to the pointer received
+in order to reach the next element in the tensor. The last argument is `int64_t n` which
+is the size of the dimension being iterated over. 
+
+The `for_each` loop will implicitly parallelize each iteration of `loop` if the size
+of each iteration is more than the value of `internal::GRAIN_SIZE`, which is a value
+that is determined as the 'right amount' of data to iterate over in order to gain a significant
+speedup using multi-threaded execution.
+
+### Setting tensor iteration dimensions
+
+The value of the strides will determine which dimension of the tensor you will iterate over.
+`TensorIterator` performs multiple optimizations internally to try to make sure that atleast
+most of the iterations happen on contiguos data to take advantage of hierarchical cache-based
+memory architectures.
```

Comment:
I am not sure I understood the very first sentence. Optimization does take place, for CPU, for example, if you put your kernel inside `cpu` folder. This folder gets compiled with special environment-specific flags for best performance.

Reply:
This is not in relation to the compilation. I am mainly talking about the optimizations that `TensorIterator` performs on a C++ level that make things like contiguous data access more probable.